lucene/core/src/java/org/apache/lucene/index/OrdinalMap.java

```diff
@@ -271,13 +273,26 @@ protected boolean lessThan(TermsEnumIndex a, TermsEnumIndex b) {
       globalOrd++;
     }
 
-    this.firstSegments = firstSegments.build();
-    this.globalOrdDeltas = globalOrdDeltas.build();
+    long ramBytesUsed = BASE_RAM_BYTES_USED + segmentMap.ramBytesUsed();
+    this.valueCount = globalOrd;
+
+    // If the first segment contains all of the global ords, then we can apply a small optimization
+    // and hardcode the first segments and global ord deltas as all zeroes.
+    if (ordDeltaBits.length > 0 && ordDeltaBits[0] == 0L && ordDeltas[0].size() == this.valueCount) {
```

Comment:
Hmm why only the first segment?  Couldn't it be the 3rd segment, in addition, that matches the global ords?

Edit: ahh OK I understand now -- this opto is indeed specific to the first segment, so we can store `this.firstSegments` as all 0s.  Good!

Do we (somewhere, couldn't find it here) pre-sort all segments by the cardinality descending?  Then we could know all segments that meet this optimization are at the start of the segments list, and possibly building the ordinal map is faster (not sure).  But then we would need to un-sort in the end to return the final `OrdinalMap`.  But it might enable this opto to apply more often, except, I think we would then need an additional dereference on lookup, hmm.

Does our `PackedLongValues.monotonicBuilder` already optimize for the case where it is all 0s, for the case where another segment (not the first) has all the global values as well?

Reply:
> Do we (somewhere, couldn't find it here) pre-sort all segments by the cardinality descending?

We do in fact -- the segments are sorted by 'weight', which in all call sites corresponds to the number of unique terms. This was added in [LUCENE-5782](https://issues.apache.org/jira/browse/LUCENE-5782).

> Does our PackedLongValues.monotonicBuilder already optimize for the case where it is all 0s, for the case where another segment (not the first) has all the global values as well?

It does look like it -- when constructing the individual `PackedInts.Reader` instances, we identify the all 0s case and use the lightweight `PackedInts.NullReader`. It's great we optimize that case, but it does mean this PR doesn't make an enormous space difference.

Reply:
> > Do we (somewhere, couldn't find it here) pre-sort all segments by the cardinality descending?
> 
> We do in fact -- the segments are sorted by 'weight', which in all call sites corresponds to the number of unique terms. This was added in [LUCENE-5782](https://issues.apache.org/jira/browse/LUCENE-5782).

Ahh, so then we know the first segment will indeed have the most unique terms, and therefore the highest chance of having "all 0s" ord deltas.

I think 2nd and 3rd segments also might have all 0s ord deltas?  But we can try to optimize that in a followon issue ... progress not perfection!

> > Does our PackedLongValues.monotonicBuilder already optimize for the case where it is all 0s, for the case where another segment (not the first) has all the global values as well?
> 
> When constructing the individual `PackedInts.Reader` instances, we do identify the all 0s case and use the lightweight `PackedInts.NullReader`. It's great we optimize that case, but it does mean this PR doesn't make an enormous space difference.

Got it.  Well, it's great that all these layers optimize :)

Reply:
> I think 2nd and 3rd segments also might have all 0s ord deltas? But we can try to optimize that in a followon issue ... progress not perfection!

I wonder if you are confused here, the proposed changes optimize the mapping from global ordinals to the ordinals of one arbitrary segment. When a segment has all value, we can simplify by always picking this segment, but there is no need to optimize this for the 2nd or 3rd segments, since we only need to be able to translate global ordinals to the ordinal of a single segment. Or maybe I'm the one confused by what you were suggesting. :)

Reply:
Aha!  Sorry, I was indeed confused ;)

This is to enable "retrieve `BytesRef` for this global ordinal" use-case, right?  For that, we first pick a segment to use (the first one also containing that `BytesRef`), then map to its segment-local ordinal, then retrieve the `BytesRef` for that using the existing doc values API for that segment.

We do not (need to, nor) expose an API today to "retrieve segment N's ordinal corresponding to global ordinal M".  Only the reverse direction (segment N's ordinal M maps to global ordinal O).

I think I understand now!

Reply:
> This is to enable "retrieve BytesRef for this global ordinal" use-case, right?

Right!

> We do not (need to, nor) expose an API today to "retrieve segment N's ordinal corresponding to global ordinal M"

Correct.